bookkeeper-tools/src/main/java/org/apache/bookkeeper/tools/cli/BookKeeperCLI.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.bookkeeper.tools.cli;
+
+import com.beust.jcommander.JCommander;
+import com.beust.jcommander.Parameter;
+import com.google.common.annotations.VisibleForTesting;
+import java.net.MalformedURLException;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.TreeMap;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.tools.cli.commands.CmdBase;
+import org.apache.bookkeeper.tools.cli.commands.CmdBookie;
+import org.apache.bookkeeper.tools.cli.commands.CmdClient;
+import org.apache.bookkeeper.tools.cli.commands.CmdCluster;
+import org.apache.bookkeeper.tools.cli.commands.CmdMetadata;
+import org.apache.commons.configuration.ConfigurationException;
+
+/**
+ * BookKeeper CLI.
+ */
+@Slf4j
+public class BookKeeperCLI {
+
+    /**
+     * Make this command map static. This provides a way to plugin different sub commands.
+     */
+    private static final Map<String, Class> commandMap;
+
+    static {
+        commandMap = new TreeMap<>();
+
+        // build the default command map
+        commandMap.put("bookie", CmdBookie.class);
+        commandMap.put("client", CmdClient.class);
+        commandMap.put("cluster", CmdCluster.class);
+        commandMap.put("metadata", CmdMetadata.class);
+    }
+
+    static JCommander newJCommander() {
+        return new JCommander();
+    }
+
+    @SuppressWarnings("unchecked")
+    @VisibleForTesting
+    public static Object newCommandInstance(Class cls, ServerConfiguration config) throws Exception {
+        return cls.getConstructor(ServerConfiguration.class).newInstance(config);
```

Comment:
any reason for using reflection? instead of instantiating directly in the static method (line 52-59)

Reply:
no particular reason for using reflection. was trying to defer object initialization when it is needed. any concern on using reflection?

Reply:
I would prefer not to use it unless there is strong need for reflection. In this case you are doing both adding commands class and instantiating them in the same class (BookKeeperCLI.java), So I dont see the need for it. Also it would be little annoying to findout from call hierarchy who are creating instances of that class and which constructor is used. Also tomorrow if we change the signature of the constructors without changing the reflection instantiation calls, reflection instantiation would fail with in runtime instead of compile time.

Reply:
@reddycharan my concern of instantiating all the commands is we basically construct all the commands, which it is completely waste resources since there is only command group is going to touched. if you check the code, we only use it for command group, for each command group we construct the sub commands. 

How about keep this for now? since it is a new CLI tool, we can decide and change what is the best solution after porting all the changes. At that point, if you still think we should not use reflection, we can change it.

Reply:
it is ok for now.